docs/code/IDataViewImplementation.md

```diff
@@ -0,0 +1,492 @@
+some everything, and it all just works, and is consistent whether we're
+applying that to the training data on which the transforms were defined, or
+some other test set, whether we programmatically load the model in the API and
+apply it to some production setting, whether we are running in a distributed
+environment and want to make sure *all* worker nodes are featurizing data in
+exactly the same way, etc. etc.
+
+The way in which this consistency is accomplished is by having certain
+requirements on the essential parts of the data model: loaders and transforms.
+The essential reason these things exist is so that they can be applied to new
+data in a consistent way.
+
+Let us formalize this somewhat. We consider two data views to be functionally
+identical if there is absolutely no way to distinguish them: they return the
+same values, have the same types, same number of rows, they shuffle
+identically given identically constructed `IRandom` when row cursors are
+constructed, return the same ID for rows from the ID getter, etc. Obviously
+this concept is transitive. (Of course, `Batch` in a cursor might be different
+between the two, but that is the case even with two cursors constructed on the
+same data view.) So some rules:
+
+1. If you have an `IDataLoader`, then saving/loading the associated data model
+   on the same data should result in a functionally identical `IDataLoader`.
+
+2. If you have an `IDataTransform`, then saving/loading the associated data
+   model for the transforms on functionally identical `IDataView`s, should
+   itself result in functionally identical `IDataView`s.
+
+## Versioning
+
+This requirement for consistency of a data model often has implications across
+versions of TLC, and our requirements for data model backwards compatibility.
+As time has passed, we often feel like it would make sense if a transform
+behaved *differently*, that is, if it organized or calculated its output in a
+different way than it currently does. For example, suppose we wanted to switch
+the hash transform to something a bit more efficient than murmur hashes, for
+example. If we did so, presumably the same input values would map to different
+outputs. We are free to do so, of course, yet: when we deserialize a hash
+transform from before we made this change, that hash transform should continue
+to output values as it did, before we made that change. (This, of course,
+assuming that the transform was released as part of a "blessed" non-preview
+point release of TLC. We can, and have, broken backwards compatibility for
+something that has not yet been incorporated in any sort of blessed release,
+though we prefer to not.)
+
+## What is Not Functionally Identical
+
+Note that identically *constructed* data views are not necessarily
+*functionally* identical. Consider this usage of the train and score transform
+with `xf=trainScore{tr=ap}`, where we first train averaged perceptron, then
```

Comment:
Recommend UpperCamel `TrainScore` as this is the case I see most often. Also TrainScore is not currently available in ML.NET. And, spelling out tr=AveragedPerceptron may be easier for readers to understand. Granted case and long vs. short is up to personal preference.

Reply:
Oooh... I thought I'd excised all `maml` command line things. This might require a bit more thought.

Unfortunately I don't know that there's a way to express this in the new public API of ML.NET though. Hmmm. :P

---
In reply to: [188811867](https://github.com/dotnet/machinelearning/pull/173#discussion_r188811867) [](ancestors = 188811867)

Reply:
Actually I just don't know what to do about this.

---
In reply to: [188843730](https://github.com/dotnet/machinelearning/pull/173#discussion_r188843730) [](ancestors = 188843730,188811867)

Reply:
Comment out the section until which point we have a TrainScore? Or perhaps leave behind in another file until we can move it back here?

Reply:
Maybe. Let me think about this.

---
In reply to: [188877540](https://github.com/dotnet/machinelearning/pull/173#discussion_r188877540) [](ancestors = 188877540)

Reply:
So I'm afraid I have to punt on this one... there's kind of no replacement for this yet. It's kind of minor, we can adjust later I hope.

---
In reply to: [189015019](https://github.com/dotnet/machinelearning/pull/173#discussion_r189015019) [](ancestors = 189015019,188877540)